### ISSUE TITLE
Net::ReadTimeout error on registering new image

### ISSUE THREAD

**Issue by kvaps** [OpenNebula/one#4920]
**Description**
Sunstone throw an error on uploading big image:

![image](https://user-images.githubusercontent.com/7556217/84408672-b8e7b780-ac0c-11ea-9f1d-95c129e7d5ce.png)

**To Reproduce*
- I was trying to upload **proxmox-ve_6.1-1.iso** (777M) but better to choose even larger image (4G)
- First it shows "Uploading... "
- Second it show "Registering in OpenNebula"
- Then throw an error:
  ```
  Cannot contact server: is it running and reachable?
  ```
  if you connecting over the reverse proxy, you will see another one:
  ```
  It appears there was a server exception. Please check server's log.
  ```
- If you refresh the page then image will successful appears. 

On browser you can see that request is not finished:
![image](https://user-images.githubusercontent.com/7556217/84409435-b5a0fb80-ac0d-11ea-9b75-48db978c29a0.png)

Packet capture shows the same:

![123_000](https://user-images.githubusercontent.com/7556217/84409682-03b5ff00-ac0e-11ea-9304-619d8708d1c2.png)

**Expected behavior**
Image successfully uploaded without any errors

**Details**
 - Affected Component: oned, sunstone
 - Hypervisor: KVM
 - Version: 5.10.5

**Additional context**
Packet capture: [sunstone_upload_bug.pcap.gz](https://github.com/OpenNebula/one/files/4765938/sunstone_upload_bug.pcap.gz)

code for `/upload`:
https://github.com/OpenNebula/one/blob/f3c50a5d898c79bb9e374eee53c0b0d8119651c0/src/sunstone/public/app/tabs/images-tab/form-panels/create-common.js#L479-L500

## Progress Status
- [ ] Branch created 
- [ ] Code committed to development branch
- [ ] Testing - QA
- [ ] Documentation 
- [ ] Release notes - resolved issues, compatibility, known issues
- [ ] Code committed to upstream release/hotfix branches
- [ ] Documentation committed to upstream release/hotfix branches

**Comment by kvaps** [comment]
Probably it is oned related problem:
```
time oneimage create /tmp/image.conf -d 198
Net::ReadTimeout

real	0m30.280s
user	0m0.215s
sys	0m0.032s
```

```bash
# cat /tmp/image.conf
NAME=test-iso
PATH=http://dl-cdn.alpinelinux.org/alpine/v3.12/releases/x86_64/alpine-standard-3.12.0-x86_64.iso
```

**Comment by kvaps** [comment]
My logs are full of these messages:

```
Fri Jun 19 16:32:08 2020 [Z0][RCM][E]: Failed to load log record at index: 468200630
Fri Jun 19 16:32:08 2020 [Z0][RCM][E]: Failed to load log record at index: 468200630
Fri Jun 19 16:32:08 2020 [Z0][RCM][E]: Failed to load log record at index: 468200630
Fri Jun 19 16:32:08 2020 [Z0][RCM][E]: Failed to load log record at index: 468200630
Fri Jun 19 16:32:08 2020 [Z0][RCM][E]: Failed to load log record at index: 468200630
Fri Jun 19 16:32:08 2020 [Z0][RCM][E]: Failed to load log record at index: 468200630
Fri Jun 19 16:32:08 2020 [Z0][RCM][E]: Failed to load log record at index: 468200630
Fri Jun 19 16:32:08 2020 [Z0][RCM][E]: Failed to load log record at index: 468200630
Fri Jun 19 16:32:08 2020 [Z0][RCM][E]: Failed to load log record at index: 468200630
Fri Jun 19 16:32:08 2020 [Z0][RCM][E]: Failed to load log record at index: 468200630
Fri Jun 19 16:32:08 2020 [Z0][RCM][E]: Failed to load log record at index: 468200630
Fri Jun 19 16:32:08 2020 [Z0][RCM][E]: Failed to load log record at index: 468200630
Fri Jun 19 16:32:08 2020 [Z0][RCM][E]: Failed to load log record at index: 468200630
Fri Jun 19 16:32:08 2020 [Z0][RCM][E]: Failed to load log record at index: 468200630
Fri Jun 19 16:32:08 2020 [Z0][RCM][E]: Failed to load log record at index: 468200630
Fri Jun 19 16:32:08 2020 [Z0][RCM][E]: Failed to load log record at index: 468200630
Fri Jun 19 16:32:08 2020 [Z0][RCM][E]: Failed to load log record at index: 468200630
Fri Jun 19 16:32:08 2020 [Z0][RCM][E]: Failed to load log record at index: 468200630
```

However RAFT it seems ok:
```
oneadmin@opennebula-oned-2:/$ onezone show 0
ZONE 0 INFORMATION                                                              
ID                : 0                   
NAME              : OpenNebula          


ZONE SERVERS                                                                    
ID NAME            ENDPOINT                                                       
 0 opennebula-oned http://opennebula-oned-0.opennebula-oned-servers:2633/RPC2
 1 opennebula-oned http://opennebula-oned-1.opennebula-oned-servers:2633/RPC2
 2 opennebula-oned http://opennebula-oned-2.opennebula-oned-servers:2633/RPC2

HA & FEDERATION SYNC STATUS                                                     
ID NAME            STATE      TERM       INDEX      COMMIT     VOTE  FED_INDEX 
 0 opennebula-oned follower   4392       473847478  473847477  -1    -1
 1 opennebula-oned follower   4392       440878502  440878502  2     -1
 2 opennebula-oned leader     4392       473847483  473847478  2     -1
```

**Comment by kvaps** [comment]
This is somehow connected with API hook:
```ini
NAME      = "iso_check_allocate"
TYPE      = api
COMMAND   = wedos/iso_check.sh
ARGUMENTS = "allocate $API"
CALL      = "one.image.allocate"
```
```bash
#!/bin/bash

# NAME      = iso_check_allocate
# TYPE      = api
# COMMAND   = iso_check.sh
# ARGUMENTS = "chtype $API"
# CALL      = "one.image.chtype"

# NAME      = iso_check_chtype
# TYPE      = api
# COMMAND   = iso_check.sh
# ARGUMENTS = "allocate $API"
# CALL      = "one.image.allocate"

ACTION="$1"
API="$2"

UTILS_PATH=/var/lib/one/remotes/datastore
XPATH="${UTILS_PATH}/xpath.rb"

unset i j XPATH_ELEMENTS

case $ACTION in
    allocate)
        while IFS= read -r -d '' element; do
            XPATH_ELEMENTS[i++]="$element"
        done < <($XPATH -b $API '/CALL_INFO/RESULT' \
                                '/CALL_INFO/EXTRA/IMAGE/ID' \
                                '/CALL_INFO/EXTRA/IMAGE/TYPE' \
                                '/CALL_INFO/PARAMETERS/PARAMETER[TYPE="IN" and POSITION=3]/VALUE')
        
        SUCCESS="${XPATH_ELEMENTS[j++]}"
        IMAGE_ID="${XPATH_ELEMENTS[j++]}"
        TYPE="${XPATH_ELEMENTS[j++]}"
        DATASTORE_ID="${XPATH_ELEMENTS[j++]}"
    ;;
    chtype)
        while IFS= read -r -d '' element; do
            XPATH_ELEMENTS[i++]="$element"
        done < <($XPATH -b $API '/CALL_INFO/RESULT' \
                                '/CALL_INFO/PARAMETERS/PARAMETER[TYPE="IN" and POSITION=2]/VALUE' \
                                '/CALL_INFO/PARAMETERS/PARAMETER[TYPE="IN" and POSITION=3]/VALUE')
        
        SUCCESS="${XPATH_ELEMENTS[j++]}"
        IMAGE_ID="${XPATH_ELEMENTS[j++]}"
        TYPE="${XPATH_ELEMENTS[j++]}"
        DATASTORE_ID=$(oneimage show -x $IMAGE_ID | $XPATH --stdin /IMAGE/DATASTORE_ID | tr -d '\0')
    ;;
    *)
        echo 'wrong method!'
        exit 1
    ;;
esac

[ "$SUCCESS" = "1" ] || exit 0

if [ "$TYPE" = "1" ] || [ "$TYPE" = "CDROM" ]; then
    EDITOR="echo LABELS=iso >" oneimage update --append $IMAGE_ID
    exit 0
fi

ISO_ONLY=($(onedatastore show -x $DATASTORE_ID | $XPATH --stdin /DATASTORE/TEMPLATE/ISO_ONLY | tr -d '\0'))

[ "$ISO_ONLY" != "yes" ] && exit 0

oneimage chtype "$IMAGE_ID" CDROM
EDITOR="echo 'LABELS=iso DEV_PREFIX=hd' >" oneimage update --append "$IMAGE_ID"
```

**Comment by kvaps** [comment]
> Fri Jun 19 16:32:08 2020 [Z0][RCM][E]: Failed to load log record at index: 468200630

Probably unrelated, after restart I still having the same issues with no messages

**Comment by kvaps** [comment]
the weird thing is that if I specify image like this:
```xml
<?xml version="1.0"?>
<methodCall>
    <methodName>one.image.allocate</methodName>
    <params>
        <param><value>oneadmin:deadbeefdeadbeefdeadbeefdeadbeef</value></param>
        <param><value>NAME=test-isoo7 SIZE=1000</value></param>
        <param><value><i4>198</i4></value></param>
    </params>
</methodCall>
```

API respond immediately, but if I use PATH instead of SIZE:

```xml
<?xml version="1.0"?>
<methodCall>
    <methodName>one.image.allocate</methodName>
    <params>
        <param><value>oneadmin:deadbeefdeadbeefdeadbeefdeadbeef</value></param>
        <param><value>NAME=test-isoo7 PATH=/var/tmp/alpine-standard-3.12.0-x86_64.iso</value></param>
        <param><value><i4>198</i4></value></param>
    </params>
</methodCall>
```

It take ages to respond

**Comment by kvaps** [comment]
`FROM_APP` is working fine, so something wrong on this stage:

https://github.com/OpenNebula/one/blob/f3c50a5d898c79bb9e374eee53c0b0d8119651c0/src/rm/RequestManagerAllocate.cc#L532

**Comment by kvaps** [comment]
@rsmontero could you provide any comment on [this](https://github.com/OpenNebula/one/issues/4920#issuecomment-647344617), please?

**Comment by kvaps** [comment]
I found the cause of the problem.
My tm/monitoring script was working quite long time every execution took about 1minut, sometimes it might hung up to 10 minutes, this behavior somehow confuses OpenNebula's Image Manager, in fact we see this behavior.

I've updated my driver to have timeouts and faster execution, but nevertheless the original bug is somewhere in OpenNebula

**Comment by kvaps** [comment]
ref https://github.com/OpenNebula/one/issues/1702

**Comment by kvaps** [comment]
It seems I was wrong, the problem is not in tm driver but in OpenNebula itself.
I seeing the same error even if my driver give an response almost immediately, but OpenNebula needs a time to parse this huge output:

```
...
VM = [ ID = 8612, POLL = "DISK_SIZE=[ID=0,SIZE=40345]"]
VM = [ ID = 8613, POLL = "DISK_SIZE=[ID=2,SIZE=292] DISK_SIZE=[ID=0,SIZE=10485]"]
VM = [ ID = 8615, POLL = "DISK_SIZE=[ID=2,SIZE=4252]"]
VM = [ ID = 8616, POLL = "DISK_SIZE=[ID=2,SIZE=3066]"]
VM = [ ID = 8617, POLL = "DISK_SIZE=[ID=3,SIZE=9524]"]
VM = [ ID = 8618, POLL = "DISK_SIZE=[ID=0,SIZE=2063]"]
...
```

right now I'm having more than 1500 such entries 🙄

**Comment by rsmontero** [comment]
So could it be that the stat operation of your TM dirver takes a very long? Note that when you create the image, once its ready to import, it is call to stat operation, this is a sync call and may be the root of the problem. I'm asking this because of it is working when you input the SIZE. Stat return the SIZE to check quotas etc...

**Comment by kvaps** [comment]
> So could it be that the stat operation of your TM dirver takes a very long?

It was my first idea, but after redesign of my driver, the execution of `tm/monitor` script took no more than 8 seconds.
I even tried to replace script to simple handle `cat /tmp/some-metrics.txt` to make it works even faster but it didn't help.

Every time when script generates all this huge output, it also slows down the image registration process.

> Note that when you create the image, once its ready to import, it is call to stat operation, this is a sync call and may be the root of the problem.

I was thinking the same but the above refutes this theory.

Now I'm thinking that `stat` is waiting until the IM driver parse all this data and put into DB.
Indeed I have no idea why it took so long time.

**Comment by jirutka** [comment]
I’m experiencing the same problem when allocating an image via `oneimage create` with URL in `--path`. It’s three years after and this bug is still not resolved.

**Comment by onenhansen** [comment]
Hello @kvaps and @jirutka ,

We have been working to resolve a similar issue although it generally has to do with the version of `curl` that is being ran which causes some of the "retry" parameters to constantly retry.  This happens because we're just grabbing the header of the file but with specific curl parameters, it considers this a failure to retry/continue the file.  This happens until the command times out (30 seconds).  

You can try to use the ` -o / --no-retry ` option when running the `oneimage create` command to avoid this issue.  You can also consider installing a lower version of `curl` to prevent that usage.   This option is available from curl version 7.71.0, so anything below that version should avoid this error as well.

If this does not resolve the issue please re-open this so we can determine an alternate cause.

https://github.com/OpenNebula/one/issues/6431
